src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,191 @@
+#include "dbgroup/lock/common.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+//                       Bit allocation of a word.
+// |     63     |       62-50       |      49-47     |        46-0       |
+// | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+namespace
+{
+/*##############################################################################
+ * Local constants
+ *############################################################################*/
+
+/// @brief An offset for right-shifting to extract the "begin position".
+constexpr uint64_t kPosShift = 47;
+
+/// @brief An offset for right-shifting to extract the "reference counter".
+constexpr uint64_t kCntShift = 50;
+
+/// @brief A constant for incrementing the "reference counter".
+constexpr uint64_t kCntUnit = 1UL << kCntShift;
+
+/// @brief A bitmask with only the "begin position" portion set to 1.
+constexpr uint64_t kPosMask = (((1UL << (1 + kCntShift - kPosShift)) - 1) << kPosShift);
+
+/// @brief A bitmask with only the "reference counter" portion set to 1.
+constexpr uint64_t kCntMask = (((1UL << (64 - kCntShift)) - 1) << kCntShift);
+
+/// @brief A bitmask with only the "descriptor address" portion set to 1.
+constexpr uint64_t kAddrMask = (1UL << 47) - 1;
+
+/// @brief A bitmask with only the "MwCAS FLAG" and "descriptor address" portions set to 1.
+constexpr uint64_t kDescMask = kMwCASFlag | kAddrMask;
+}  // namespace
+
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  auto *page = tls.release();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;
+  return desc;
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      if (!EmbedDescriptor(desc_addr, i)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  auto ret = (cur_stat == kSucceeded);
```

Comment:
`ret` だと成功したのかがわかりにくいので， `succeeded` の方がよいと思います．また，実質定数なので `const` で．

```suggestion
  const auto succeeded = (cur_stat == kSucceeded);
```